paasta_tools/paasta_maintenance.py

```diff
@@ -0,0 +1,276 @@
+        'action',
+        choices=['drain', 'undrain', 'down', 'up', 'status'],
+        help="Action to perform on the speicifed hosts",
+    )
+    parser.add_argument(
+        'hostname',
+        nargs='*',
+        help="Hostname(s) of machine(s) to start draining.",
+    )
+    options = parser.parse_args()
+    return options
+
+
+def base_api():
+    leader = get_mesos_leader()
+
+    def execute_request(method, endpoint, **kwargs):
+        url = "http://%s:%d/%s" % (leader, PORT, endpoint)
+        timeout = 15
+        s = Session()
+        req = Request(method, url, **kwargs)
+        prepared = s.prepare_request(req)
+        try:
+            resp = s.send(prepared,
+                          timeout=timeout,
+                          auth=load_credentials(),
+                          )
+            resp.raise_for_status()
+        except Exception as e:
+            print("Error executing API request calling %s. Got response code %d with error %s" %
+                  (endpoint, resp.status_code, e))
+    return execute_request
+
+
+def master_api():
+    def execute_master_api_request(method, endpoint, **kwargs):
+        base_api_client = base_api()
+        return base_api_client(method, "master/%s" % endpoint, **kwargs)
+    return execute_master_api_request
+
+
+def maintenance_api():
+    def execute_schedule_api_request(method, endpoint, **kwargs):
+        master_api_client = master_api()
+        return master_api_client(method, "maintenance/%s" % endpoint, **kwargs)
+    return execute_schedule_api_request
+
+
+def get_schedule_client():
+    def execute_schedule_api_request(method, endpoint, **kwargs):
+        maintenance_api_client = maintenance_api()
+        return maintenance_api_client(method, "schedule/%s" % endpoint, **kwargs)
+    return execute_schedule_api_request
+
+
+def get_maintenance_schedule():
+    client_fn = get_schedule_client()
+    return client_fn(method="GET", endpoint="")
+
+
+def get_maintenance_status():
+    client_fn = get_schedule_client()
+    return client_fn(method="GET", endpoint="status")
+
+
+def timedelta_type(value):
+    """Return the delta in nanoseconds.
+    :param value: a string containing a time format supported by :mod:`pytimeparse`
+    """
+    if value is None:
+        return None
+    return seconds_to_nanoseconds(timeparse.timeparse(value))
+
+
+def datetime_seconds_from_now(seconds):
+    return now() + datetime.timedelta(seconds=seconds)
+
+
+def now():
+    return datetime.datetime.now(tz.tzutc())
+
+
+def seconds_to_nanoseconds(seconds):
+    return seconds * 1000000000
+
+
+def datetime_to_nanoseconds(dt):
+    return seconds_to_nanoseconds(int(dt.strftime("%s")))
+
+
+def build_start_maintenance_payload(hostnames):
+    return get_machine_ids(hostnames)
+
+
+def get_machine_ids(hostnames):
+    machine_ids = []
+    for hostname in hostnames:
+        machine_id = dict()
+        # This is to allow specifying a hostname as "hostname|ipaddr"a
```

Comment:
I'm not worried at all about the overhead. We ran into a problem a while ago where we updated DNS and I marked a box as being 'down' in preparation for it being imaged. However, there was some error with the DNS change that required us to change where the hostname pointed. Once the revised DNS change was pushed, I had no easy way to mark the old host/ip mapping as being up (since the maintenance script was resolving the hostname to a different IP).

The other case where this is useful is if you want to mark a box as 'down' prior to it even existing in DNS.